packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -299,34 +401,67 @@ export function resetHooks(): void {
   if (__DEV__) {
     currentHookNameInDev = undefined;
   }
-
   didScheduleRenderPhaseUpdate = false;
   renderPhaseUpdates = null;
   numberOfReRenders = -1;
 }
 
 function createHook(): Hook {
-  return {
-    memoizedState: null,
+  let hook: Hook = __DEV__
+    ? {
+        _debugType: ((currentHookType: any): HookType),
+        memoizedState: null,
 
-    baseState: null,
-    queue: null,
-    baseUpdate: null,
+        baseState: null,
+        queue: null,
+        baseUpdate: null,
 
-    next: null,
-  };
+        next: null,
+      }
+    : {
+        memoizedState: null,
+
+        baseState: null,
+        queue: null,
+        baseUpdate: null,
+
+        next: null,
+      };
+
+  return hook;
 }
 
 function cloneHook(hook: Hook): Hook {
-  return {
-    memoizedState: hook.memoizedState,
+  let nextHook: Hook = __DEV__
+    ? {
+        _debugType: ((currentHookType: any): HookType),
+        memoizedState: hook.memoizedState,
 
-    baseState: hook.baseState,
-    queue: hook.queue,
-    baseUpdate: hook.baseUpdate,
+        baseState: hook.baseState,
+        queue: hook.queue,
+        baseUpdate: hook.baseUpdate,
 
-    next: null,
-  };
+        next: null,
+      }
+    : {
+        memoizedState: hook.memoizedState,
+
+        baseState: hook.baseState,
+        queue: hook.queue,
+        baseUpdate: hook.baseUpdate,
+
+        next: null,
+      };
+
+  if (__DEV__ && !currentHookMismatch) {
```

Comment:
Nit: we try to leave `if (__DEV__)` as its own condition for it to jump out — and nest `if`s

Reply:
missed this